### ISSUE TITLE
android package name starting with a kotlin keyword

### ISSUE THREAD

**Issue by pvinis** [expo/expo#26910]
### Summary

When a package name starts with a kotlin keyword, like `is` or `in`, android cant build, with an error of `Package name must be a '.'-separated identifier list`.

`.is` for example is the tld for Iceland, so a website like `example.is` making an app, would be something like `is.example.mobileapp`.

That translates to a kotlin file that starts with `package is.example.mobileapp` and gradle cant build.

According to https://stackoverflow.com/a/47265334, the solution is to wrap the keyword in backticks, so it would become ```package `is`.example.mobileapp```

### What platform(s) does this occur on?

Android

### SDK Version

50

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 14.3
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.18.0 - ~/.local/share/mise/installs/node/16.18.0/bin/node
      Yarn: 1.22.19 - ~/.local/share/mise/installs/yarn/1.22.19/bin/yarn
      npm: 8.19.2 - ~/.local/share/mise/installs/node/16.18.0/bin/npm
      Watchman: 2023.12.04.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
    IDEs:
      Android Studio: 2023.1 AI-231.9392.1.2311.11076708
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmPackages:
      expo: ^50.0.5 => 50.0.5
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.73.2 => 0.73.2
    Expo Workflow: bare

### Minimal reproducible example

app.json with 
```
	android: {
		package: "is.pvin.crossings",
```

**Comment by linonetwo** [comment]
use `ren.onetwo.mobile` instead of `fun.tidgi.mobile` fix this

**Comment by saravanakumargn** [comment]
``` java
package `in`.pvin.crossings
```

**Comment by akshitfudr** [comment]
Any solution to this, as I'm unable to create android builds. The package name starts with "in".

**Comment by chendoron** [comment]
Please prioritize this, it's blocking us from upgrading to SDK 50.

**Comment by Kaddtechnologies** [comment]
This is still an issue. I need to use the word 'In' as this is a requirement

**Comment by informsharique** [comment]
@brentvatne @EvanBacon Please look into this.

**Comment by srkgupta** [comment]
Hi, 

I was able to fix  this locally and apply a local patch through the following steps. 

TLDR: It replaces the conflicting keywords within ` characters. For example: in.package.name with `in`.package.name 


- Install patch-package:

`npm install patch-package --save-dev  # or yarn add patch-package --dev`

- Modify the desired file within the node_modules/@expo/config-plugins/build/android/Package.js

After the line:

```js
let contents = _fs().default.readFileSync(filepath).toString();
```

Replace the line:

```js
contents = replacePackageName(contents, currentPackageName, packageName);
```

With:

```js
if (_path().default.extname(filepath) === ".kt") {
  contents = replacePackageName(
    contents,
    currentPackageName,
    kotlinSanitized(packageName)
  );
} else {
  contents = replacePackageName(contents, currentPackageName, packageName);
}
```

After the function, add:

```js
function kotlinSanitized(packageName) {
  const stringsToWrap = ["is", "in", "as", "fun"];
  const parts = packageName.split(".");
  const cleanParts = parts.map((part) =>
    stringsToWrap.includes(part) ? "`" + part + "`" : part
  );

  const cleanName = cleanParts.join(".");
  return cleanName;
}
```

- Run `npx patch-package @expo/config-plugins` to generate a patch file. This will create a .patch file in a newly created patches folder within your project root.
- Update your build script (e.g., package.json scripts section) to execute patch-package before the build process. Here's an example using npm script:

```json
{
  "scripts": {
    "postinstall": "patch-package"
  }
}
```

Here's the patch file it generated. 
[@expo+config-plugins+7.9.1.patch](https://github.com/expo/expo/files/15152184/%40expo%2Bconfig-plugins%2B7.9.1.patch)

**Comment by lukmccall** [comment]
Closing because it was fixed by https://github.com/expo/expo/pull/27458. The fix should be included in SDK 51.

**Comment by detanx** [comment]
@pvinis I still encountered this problem when using expo51. Can you take a look at the issue I raised? Thank you.  [#28821](https://github.com/expo/expo/issues/28821)

**Comment by pvinis** [comment]
I see your package name is https://github.com/detanx/expo-react-native-bluetooth-escpos-printer/blob/main/app.json#L23, so I dont see how this is connected.

**Comment by robiulm60ict** [comment]
This Error 
Launching lib\main.dart on sdk gphone64 x86 64 in debug mode...
Running Gradle task 'assembleDebug'...
e: C:/Users/M360 App 1/.gradle/caches/transforms-3/884371f8ba4c22f7d845ae45913c9f76/transformed/jetified-kotlin-stdlib-2.0.0.jar!/META-INF/kotlin-stdlib-jdk7.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.7.1.
e: C:/Users/M360 App 1/.gradle/caches/transforms-3/884371f8ba4c22f7d845ae45913c9f76/transformed/jetified-kotlin-stdlib-2.0.0.jar!/META-INF/kotlin-stdlib-jdk8.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.7.1.
e: C:/Users/M360 App 1/.gradle/caches/transforms-3/884371f8ba4c22f7d845ae45913c9f76/transformed/jetified-kotlin-stdlib-2.0.0.jar!/META-INF/kotlin-stdlib.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.7.1.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 9s

┌─ Flutter Fix ──────────────────────────────────────────────────────────────────────────────┐
│ [!] Your project requires a newer version of the Kotlin Gradle plugin.                     │
│ Find the latest version on https://kotlinlang.org/docs/releases.html#release-details, then │
│ update G:\RobiulFlutterProject\crm-zayed_robiul\android\build.gradle:                      │
│ ext.kotlin_version = '<latest-version>'                                                    │
└────────────────────────────────────────────────────────────────────────────────────────────┘
Error: Gradle task assembleDebug failed with exit code 1

packace version:
  geolocator: ^12.0.0
  geolocator_android: ^4.5.5



buildscript {
    ext.kotlin_version = '2.0.0'
    repositories {
        google()
        mavenCentral()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:7.4.2'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"


    }
}

allprojects {
    repositories {
        google()
        mavenCentral()
    }
}

rootProject.buildDir = "../build"
subprojects {
    project.buildDir = "${rootProject.buildDir}/${project.name}"
}
subprojects {
    project.evaluationDependsOn(":app")
}

tasks.register("clean", Delete) {
    delete rootProject.buildDir
}

plugins {
    id "com.android.application"
    // START: FlutterFire Configuration
    id 'com.google.gms.google-services'
    // END: FlutterFire Configuration
    id "kotlin-android"
    // The Flutter Gradle Plugin must be applied after the Android and Kotlin Gradle plugins.
    id "dev.flutter.flutter-gradle-plugin"

}

def localProperties = new Properties()
def localPropertiesFile = rootProject.file("local.properties")
if (localPropertiesFile.exists()) {
    localPropertiesFile.withReader("UTF-8") { reader ->
        localProperties.load(reader)
    }
}

def flutterVersionCode = localProperties.getProperty("flutter.versionCode")
if (flutterVersionCode == null) {
    flutterVersionCode = "1"
}

def flutterVersionName = localProperties.getProperty("flutter.versionName")
if (flutterVersionName == null) {
    flutterVersionName = "1.0"
}

def keystoreProperties = new Properties()
def keystorePropertiesFile = rootProject.file('key.properties')
if (keystorePropertiesFile.exists()) {
    keystoreProperties.load(new FileInputStream(keystorePropertiesFile))
}
android {
    namespace = "com.crm360employee.crm"
    compileSdk = 34
    ndkVersion "25.1.8937393"

    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_1_8
        targetCompatibility = JavaVersion.VERSION_1_8
    }

    defaultConfig {

        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId = "com.crm360employee.crm"
        // You can update the following values to match your application needs.
        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-gradle-build-configuration.
        minSdk = flutter.minSdkVersion
        targetSdk = flutter.targetSdkVersion
        versionCode = flutterVersionCode.toInteger()
        versionName = flutterVersionName

    }

    signingConfigs {
        release {
            keyAlias keystoreProperties['keyAlias']
            keyPassword keystoreProperties['keyPassword']
            storeFile keystoreProperties['storeFile'] ? file(keystoreProperties['storeFile']) : null
            storePassword keystoreProperties['storePassword']
        }
    }
    buildTypes {
        release {
            // TODO: Add your own signing config for the release build.
            // Signing with the debug keys for now,
            // so `flutter run --release` works.
//                      signingConfig signingConfigs.debug
            signingConfig signingConfigs.release
        }
    }
}
dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    // Other dependencies
}

flutter {
    source = "../.."
}
  cupertino_icons: ^1.0.6
  page_transition:
  intl:
  url_launcher:
  get:
  shared_preferences:
  http:
  file_picker: ^8.0.3
  image_picker: ^1.1.2
  lottie:
  in_app_update:
  rive: ^0.13.5
  iconsax_flutter: ^1.0.0
  syncfusion_flutter_calendar: ^26.1.41
  cached_network_image: ^3.3.1
  shimmer: ^3.0.0
  iconsax: ^0.0.8
  path_provider: ^2.1.3
  flutter_image_compress: ^2.3.0
  gallery_saver: ^2.1.1
  percent_indicator: ^4.2.3
  flutter_otp_text_field: ^1.1.3+2
  device_info_plus: ^10.1.0
  firebase_core: ^3.2.0
  firebase_messaging: ^15.0.3
  flutter_local_notifications: ^17.2.1+1
  animation_list: ^3.0.0
  flutter_background_service: ^5.0.6
  geolocator: ^12.0.0
  geolocator_android: ^4.5.5
  connectivity_plus: ^6.0.3
  sqflite: ^2.3.3+1
  socket_io_client: ^3.0.0-beta.2
  googleapis:
  googleapis_auth:
  geocoding: ^3.0.0
  badges: ^3.1.2
  primer_progress_bar: ^0.4.2
  aura_box: ^1.0.2

**Comment by wodin** [comment]
@robiulm60ict 

> Flutter Fix

You're using Flutter. Not Expo or React Native.

If you want to try writing your app in React Native, you can start here:
https://docs.expo.dev/tutorial/introduction/

If you want to stick with Flutter, you'll need to find a Flutter forum to ask your question on

**Comment by faustoct1** [comment]
installed fresh new expo/react-native problem still happening.
